server/src/main/java/org/elasticsearch/cluster/coordination/StableMasterHealthIndicatorService.java

```diff
@@ -0,0 +1,208 @@
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.discovery.DiscoveryModule;
+import org.elasticsearch.health.HealthIndicatorDetails;
+import org.elasticsearch.health.HealthIndicatorImpact;
+import org.elasticsearch.health.HealthIndicatorResult;
+import org.elasticsearch.health.HealthIndicatorService;
+import org.elasticsearch.health.HealthStatus;
+import org.elasticsearch.health.ImpactArea;
+import org.elasticsearch.health.SimpleHealthIndicatorDetails;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.elasticsearch.health.ServerHealthComponents.CLUSTER_COORDINATION;
+
+public class StableMasterHealthIndicatorService implements HealthIndicatorService, ClusterStateListener {
+
+    public static final String NAME = "stable_master";
+
+    private final ClusterService clusterService;
+    private final DiscoveryModule discoveryModule;
+    private final MasterHistoryService masterHistoryService;
+    private static final Logger logger = LogManager.getLogger(StableMasterHealthIndicatorService.class);
+
+    public StableMasterHealthIndicatorService(
+        ClusterService clusterService,
+        DiscoveryModule discoveryModule,
+        MasterHistoryService masterHistoryService
+    ) {
+        this.clusterService = clusterService;
+        this.discoveryModule = discoveryModule;
+        this.masterHistoryService = masterHistoryService;
+        clusterService.addListener(this);
+    }
+
+    @Override
+    public String name() {
+        return NAME;
+    }
+
+    @Override
+    public String component() {
+        return CLUSTER_COORDINATION;
+    }
+
+    @Override
+    public HealthIndicatorResult calculate(boolean includeDetails) {
+        HealthStatus stableMasterStatus;
+        String summary;
+        Map<String, Object> details = new HashMap<>();
+        Collection<HealthIndicatorImpact> impacts = new ArrayList<>();
+        MasterHistory localMasterHistory = masterHistoryService.getLocalMasterHistory();
+        if (hasSeenMasterInLast30Seconds()) {
+            long masterChanges = getNumberOfMasterChanges(localMasterHistory);
+            logger.trace("Have seen a master in the last 30 seconds");
+            if (localMasterHistory.hasSameMasterGoneNullNTimes(4)) {
+                DiscoveryNode master = localMasterHistory.getMostRecentNonNullMaster();
+                logger.trace("One master has gone null 4 or more times recently: " + master);
+                boolean localNodeIsMaster = clusterService.localNode().equals(master);
+                List<DiscoveryNode> remoteHistory = localNodeIsMaster
```

Comment:
This is rather confusing. This seems to be renaming the local history to `remoteHistory`? I believe it's the remote history we want to check - if we are the master node then all is green and we're done here no?

Reply:
If we are the master then I think the status is yellow, b/c we've confirmed that it's flopped between null and not nulll at least 4 times. This code was an attempt to not complicate the logic in the event that the local node is master, but it apparently did not succeed. I'll make it more explicit.